jellyfin_kodi/downloader.py

```diff
@@ -279,29 +279,51 @@ def get_query_params(params, start, count):
         # complete all tasks before allowing any results to be processed. ThreadPoolExecutor
         # allows for completed tasks to be processed while other tasks are completed on other
         # threads. Dont be a dummy.Pool, be a ThreadPoolExecutor
-        p = concurrent.futures.ThreadPoolExecutor(DTHREADS)
-
-        results = p.map(lambda params: _get(url, params, server_id=server_id), query_params)
-
-        for params, result in zip(query_params, results):
-            query['params'] = params
-
-            result = result or {'Items': []}
-
-            # Mitigates #216 till the server validates the date provided is valid
-            if result['Items'][0].get('ProductionYear'):
-                try:
-                    date(result['Items'][0]['ProductionYear'], 1, 1)
-                except ValueError:
-                    LOG.info('#216 mitigation triggered. Setting ProductionYear to None')
-                    result['Items'][0]['ProductionYear'] = None
-
-            items['Items'].extend(result['Items'])
-            # Using items to return data and communicate a restore point back to the callee is
-            # a violation of the SRP. TODO: Seperate responsibilities.
-            items['RestorePoint'] = query
-            yield items
-            del items['Items'][:]
+        with concurrent.futures.ThreadPoolExecutor(DTHREADS) as p:
+            # dictionary for storing the jobs and their results
+            jobs = {}
+
+            # semaphore to avoid fetching complete library to memory
+            thread_buffer = threading.Semaphore(LIMIT * DTHREADS)
+
+            # wrapper function for _get that uses a semaphore
+            def get_wrapper(params):
+                thread_buffer.acquire()
+                return _get(url, params, server_id=server_id)
```

Comment:
Random thought I just had, but I'm not currently able to test/check for myself.

Is this setting the max _items_ returns from the api, or the max _groups_?  `_get()` is returning a json object that contains `limit` number of items inside of it.

If it's the max items and it's crashing at 150 items, that's crazy.  If it's max groups, then we're essentially trying to store `limit * dthreads` json objects containing `limit` items, which will add up really quick.  Going off the defaults, that's 3 * 50 * 50 = 7500 items in memory.

Reply:
This problem was already discussed above, see comments https://github.com/jellyfin/jellyfin-kodi/pull/410#issuecomment-709600405 and https://github.com/jellyfin/jellyfin-kodi/pull/410#issuecomment-709627868

As mentioned in https://github.com/jellyfin/jellyfin-kodi/pull/410#issuecomment-713060608 the semaphore size should be set to `DTHREADS`:
```suggestion
            # semaphore to avoid fetching complete library to memory
            thread_buffer = threading.Semaphore(DTHREADS)

            # wrapper function for _get that uses a semaphore
            def get_wrapper(params):
                thread_buffer.acquire()
                return _get(url, params, server_id=server_id)
```

If this will be implemented, maybe a better way will be using a threading queue instead of a semaphore as suggested by @oddstr13

Reply:
I think with this change it does appear to fix the problem.  I finally managed to reproduce the memory exhaustion on one of my Pis, but after implementing this fix and changing it to `thread_buffer = threading.Semaphore(DTHREADS)` it allowed the scan to finish and only hovered around 25% memory usage.

Reply:
Ok, applied that change